File: 0094 Binary Tree Inorder Traversal/solution.cpp
#include <bits/stdc++.h>
using namespace std;

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 };

class Solution {
public:
    vector<int> inorderTraversal(TreeNode* root) 
    {
        stack<TreeNode *> s;
        vector<int> res;
        
        while (root || s.size())
        {
            while(root)
            {
                s.push(root);
                root = root->left;
            }
            root = s.top(), s.pop();
            res.push_back(root->val);
            
            root = root->right;
        }
        return res;
    }
    
};

int main(int argc, char** argv) {
  TreeNode* root = new TreeNode(1);
  TreeNode* node2 = new TreeNode(2);
  TreeNode* node3 = new TreeNode(3);
    
  root -> right = node2;
  node2 -> left = node3;
    
  Solution s;
  for (auto i : s.inorderTraversal(root))
    std::cout << i << " ";
  std::cout << std::endl; 
    
  return 0;
}

File: 0094 Binary Tree Inorder Traversal/solution.py
#! python3

class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class Solution:
    def inorderTraversal(self, root):
        """
        :type root: TreeNode
        :rtype: List[int]
        """
        # Recursive
        res = []
        self.helper(root, res)
        return res
        
    def helper(self, root, res):
        if not root: return ;
        self.helper(root.left, res)
        res.append(root.val)
        self.helper(root.right, res)

if __name__ == "__main__":
    root, node2, node3= TreeNode(1), TreeNode(2), TreeNode(3)
    root.right, node2.left = node2, node3

    print(Solution().inorderTraversal(root))